### ISSUE TITLE
Обход запрета на загрузку аудио

### ISSUE THREAD

**Issue by Pmmlabs** [VkOpt/VkOpt#257]
Вконтакте блокирует загрузку аудиозаписей, которые были изъяты по просьбе правообладателей. Предлагаю добавить функцию обхода этой блокировки в Вкопт:
![default](https://cloud.githubusercontent.com/assets/2682026/11763271/a996582a-a114-11e5-86ac-6ba4a8969528.png)
Для идентификации песен используется "цифровой слепок", но он не идеален. Известно, что существуют методы обхода: раньше например я добавлял две секунды синусоиды 20 Гц в начало и запись успешно загружалась. Сегодня я обнаружил, что этот способ перестал работать, однако продолжает работать метод незначительной эквализации: достаточно опустить уровень сигнала в диапазоне 200-300 Гц на 5 дБ, и песня загружается успешно, при этом качество не ухудшается, а провал басов почти не ощутим, так сказать, невооруженным ухом. (в наушниках-затычках Sony я разницы не ощутил)
График эквализации:
![default](https://cloud.githubusercontent.com/assets/2682026/11763492/89e41e64-a11d-11e5-9dc0-2afdd5a1b6d3.png)

При установлении галочки будет происходить предварительная обработка аудиозаписи, после которой пойдет непосредственно загрузка. Обработка - это понижение уровня некоторого диапазона частот. Для декодирования аудио можно задействовать Web Audio API, как [сделано здесь](http://habrahabr.ru/post/240819/). Правда я пока не знаю, можно ли обработанный таким образом звук запихать обратно в <audio> чтобы получить данные для отправки.
В общем, кто что думает по этому поводу? По-моему, идея довольно реальная.

**Comment by Corak667** [comment]
Идея хорошая, я только за. Но у меня уже разработан свой метод, более гуманный:
Зачем что-то менять в песне? Можно и без потерь - я для этого использую метод без потерь - просто в начало трека через mp3directcut - он режет+склеивает и редактирует уровни усиления по блокам mp3 без пересжатия. Через него добавляю В НАЧАЛО рендомные нарезки фрагмента трека приглушенные на -79 дб (тишину) - от 15 до 100 секунд - в зависимости от ключевого фрагмента слепка. 
У их анализатора слепка - лимит по времени, отодвинув время ключевого момента трека, анализатор ничего не сможет сделать и пропустит трек. Единственный минус - придется при прослушивании скипать тишину в этом диапазоне времени, но зато сохранится аутентичный 1:1 по качеству и звуку. Ухудшать как-то трек не вижу смысла. Хотя твой вариант тоже хорош, но только в случае если он не сильно портит трек. Лучше такие треки обозначать отдельно...
 Также существует еще много методов - обрезать трек на пару частей, чтобы обойти тот ключевой фрагмент. Анализ слепка идет от начала трека - добавление тишины вконец не спасет. Также можно склеить несколько треков, например - заливать альбом целиком, тогда искомый трек будет в середине и система анализа его тоже скипнет...

**Comment by Pmmlabs** [comment]
@Corak667, не, это всё хуже. Вот скину я человеку песню, а он послушает 3 секунды и напишет "ты что мне тишину скинул"? вот раньше было нормально, когда можно было 2 секунды шума в начало вставить. А если сейчас нужно от 15 до 100 секунд вставлять - то это вообще никуда не годится. Альбом целиком тоже плохо, поиск не находит и потом трудно искать внутри альбома песню.
Кстати, я сейчас проверил, можно тот диапазон частот не опускать. а наоборот поднимать - тогда это даже получится не потеря, а улучшение.

**Comment by Pmmlabs** [comment]
> Так что мое предложение проще - добавлять в начало трека любые его мусорные фрагменты с -79дб усилением (чтобы были идентичны битрейт и остальные параметры mp3 файла). Динамически загружать и проверять от 15 до 100 секунд тишины по 10 секунд, пока не обойдет ограничения и не загрузится окончательно. 

Да, это проще, но загружать файл несколько раз - это очень долго, плюс увеличивается размер самого файла, плюс тишина в начале - это крайне неудобно, я хочу нажать кнопку и чтобы пошла музыка, а не сидеть две минуты ждать, когда же она начнется. Честно говоря, бесят вот такие аудио, склееные, разбитые, смещенные. Мне пофиг на байты один в один, моё ухо это не слышит, главное реальное качество остаётся.

> Если потребуется пересжатие mp3 - у меня есть аудиофильская - практически lossless схема сжатия, без потерь. Через Lame3.93.1 с правильным пресетом.

Спасибо, но больше пригодятся наверное не аргументы командной строки, а код вызова библиотечных функций Lame, потому что если будет использоваться пересжатие, то будет использоваться библиотека Lame, портированная на Javascript, то есть там будет не запускаемый бинарник, а библиотека, предоставляющая API. Хотя по мануалу наверное можно будет перевести.

**Comment by Pmmlabs** [comment]
Сегодня разобрался, как [декодировать mp3 на JS](https://github.com/audiocogs/jsmad) и [обрабатывать](https://github.com/jussi-kalliokoski/audiolib.js) чистые PCM данные. Осталось освоить Lame и будет готово.
По производительности - четырехминутная песенка декодируется где-то за 10 секунд, это не намного больше, чем в нативных приложениях.
По частотному спектру - есть небольшие потери в самых верхах. Сохранял wav из браузера и из GoldWave, сравнивал через Spek ([скриншот](https://cloud.githubusercontent.com/assets/2682026/11920008/63607948-a773-11e5-9b54-30e42c11c093.png), слева GoldWave, справа JS)

**Comment by Corak667** [comment]
@Pmmlabs Ну это совсем нестрашные потери, которые даже делают трек чище. Оно как раз срезает аномальный редкий шум оставшийся от предыдущего урезания. Главное, что тут остаются самые громкие элементы звука.
